src/java.base/share/classes/java/util/concurrent/StructuredTaskScopeImpl.java

```diff
@@ -0,0 +1,442 @@
+            scope.onComplete(this);
+        }
+
+        @Override
+        public Subtask.State state() {
+            Object result = this.result;
+            if (result == null) {
+                return State.UNAVAILABLE;
+            } else if (result instanceof AltResult alt) {
+                // null or failed
+                return alt.state();
+            } else {
+                return State.SUCCESS;
+            }
+        }
+
+        @Override
+        public T get() {
+            scope.ensureJoinedIfOwner();
+            Object result = this.result;
+            if (result instanceof AltResult) {
+                if (result == RESULT_NULL) return null;
+            } else if (result != null) {
+                @SuppressWarnings("unchecked")
+                T r = (T) result;
+                return r;
+            }
+            throw new IllegalStateException(
+                    "Result is unavailable or subtask did not complete successfully");
+        }
+
+        @Override
+        public Throwable exception() {
+            scope.ensureJoinedIfOwner();
+            Object result = this.result;
+            if (result instanceof AltResult alt && alt.state() == State.FAILED) {
+                return alt.exception();
+            }
+            throw new IllegalStateException(
+                    "Exception is unavailable or subtask did not complete with exception");
+        }
+
+        @Override
+        public String toString() {
+            String stateAsString = switch (state()) {
+                case UNAVAILABLE -> "[Unavailable]";
+                case SUCCESS     -> "[Completed successfully]";
+                case FAILED      -> {
+                    Throwable ex = ((AltResult) result).exception();
+                    yield "[Failed: " + ex + "]";
+                }
+            };
+            return Objects.toIdentityString(this) + stateAsString;
+        }
+    }
+
+    /**
+     * Configuration implementation.
+     */
+    record ConfigImpl(ThreadFactory threadFactory,
+                      String name,
+                      Duration timeout) implements Configuration {
+        static Configuration defaultConfig() {
+            return new ConfigImpl(Thread.ofVirtual().factory(), null, null);
+        }
+
+        @Override
+        public Configuration withThreadFactory(ThreadFactory threadFactory) {
+            return new ConfigImpl(Objects.requireNonNull(threadFactory), name, timeout);
+        }
+
+        @Override
+        public Configuration withName(String name) {
+            return new ConfigImpl(threadFactory, Objects.requireNonNull(name), timeout);
+        }
+
+        @Override
+        public Configuration withTimeout(Duration timeout) {
+            return new ConfigImpl(threadFactory, name, Objects.requireNonNull(timeout));
+        }
+    }
+
+    /**
+     * Used to schedule a task to cancel the scope when a timeout expires.
+     */
+    private static class TimerSupport {
```

Comment:
Might make sense to use the commonPool():s DelayScheduler?

Reply:
TimerSupport hasn't changed, and pre-dates the FJP work to support delayed tasks, so maybe now is the time to migrate it.